### DISCUSSION TITLE
Reliable tests with 6.2 RavenDB.TestDriver?

### DISCUSSION THREAD

**Discussion by SbiCA** [ravendb/ravendb#19412] [Q&A] [answered]
Hi, I recently updated from `6.0.105` to `6.2`.

I got an following exception (which means someone thought about the test case ❤️) about missing license

after exploring the options I saw one can use the TestServer options to disable `ThrowOnInvalidOrMissingLicense` but that is randomly (~ 100 of 700 tests fail on my mac on, CI linux it's different ...)  failing  here is what I tried 

```c#

using Raven.Client.Documents;
using Raven.Embedded;
using Raven.TestDriver;
using SamSaidYes.Infrastructure.RavenDb;

namespace SamSaidYes.PostsService.Tests.TestHelpers;

public class BaseRavenTestDriver : RavenTestDriver
{
    // Option 1
    // static BaseRavenTestDriver()
    // {
    //     ConfigureServer(new TestServerOptions
    //     {
    //         Licensing = new ServerOptions.LicensingOptions
    //         {
    //             ThrowOnInvalidOrMissingLicense = false
    //         }
    //     });
    // }

    protected BaseRavenTestDriver(int? maxNumberOfRequest = null)
    {
        MaxNumberOfRequests = maxNumberOfRequest;     
        // Option 2
        // ConfigureServer(new TestServerOptions
        // {
        //     Licensing = new ServerOptions.LicensingOptions
        //     {
        //         ThrowOnInvalidOrMissingLicense = false
        //     }
        // });
    }

    public int? MaxNumberOfRequests { get; set; }

    protected override void PreInitialize(IDocumentStore documentStore)
    {
        documentStore.ApplyDocumentConventions(Setup.RavenConventions);
        if (MaxNumberOfRequests is not null)
        {
            documentStore.Conventions.MaxNumberOfRequestsPerSession = MaxNumberOfRequests.Value;
        }

        documentStore.Conventions.DisableTcpCompression = true;
    }
}

```

unfortunately neither setting them in a static constructor of the `Base` 

``` c#
System.TypeInitializationException: The type initializer for 'SamSaidYes.PostsService.Tests.TestHelpers.BaseRavenTestDriver' threw an exception.

System.TypeInitializationException
The type initializer for 'SamSaidYes.PostsService.Tests.TestHelpers.BaseRavenTestDriver' threw an exception.
   at SamSaidYes.PostsService.Tests.TestHelpers.BaseRavenTestDriver..ctor(Nullable`1 maxNumberOfRequest) in /Users/simon/Documents/git/ssy/Backend/SamSaidYes.PostsService.Tests/TestHelpers/BaseRavenTestDriver.cs:line 22
   at SamSaidYes.PostsService.Tests.ApiTest.Analytics.LeaderBoardServiceTest..ctor(ITestOutputHelper outputHelper) in /Users/simon/Documents/git/ssy/Backend/SamSaidYes.PostsService.Tests/ApiTest/Analytics/LeaderBoardServiceTest.cs:line 25
   at InvokeStub_LeaderBoardServiceTest..ctor(Object, Span`1)
   at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)

System.InvalidOperationException
Cannot configure server after it was started. Please call 'ConfigureServer' method before any 'GetDocumentStore' is called.
   at Raven.TestDriver.RavenTestDriver.ConfigureServer(TestServerOptions options) in /_/62001/src/Raven.TestDriver/RavenTestDriver.cs:line 63
   at SamSaidYes.PostsService.Tests.TestHelpers.BaseRavenTestDriver..cctor() in /Users/simon/Documents/git/ssy/Backend/SamSaidYes.PostsService.Tests/TestHelpers/BaseRavenTestDriver.cs:line 13
```

nor using the immediate class constructor of the test class solves this

``` c#
System.InvalidOperationException: Cannot configure server after it was started. Please call 'ConfigureServer' method before any 'GetDocumentStore' is ...

System.InvalidOperationException
Cannot configure server after it was started. Please call 'ConfigureServer' method before any 'GetDocumentStore' is called.
   at Raven.TestDriver.RavenTestDriver.ConfigureServer(TestServerOptions options) in /_/62001/src/Raven.TestDriver/RavenTestDriver.cs:line 63
   at SamSaidYes.PostsService.Tests.TestHelpers.BaseRavenTestDriver..ctor(Nullable`1 maxNumberOfRequest) in /Users/simon/Documents/git/ssy/Backend/SamSaidYes.PostsService.Tests/TestHelpers/BaseRavenTestDriver.cs:line 13
   at SamSaidYes.PostsService.Tests.ApiTest.Analytics.LeaderBoardServiceTest..ctor(ITestOutputHelper outputHelper) in /Users/simon/Documents/git/ssy/Backend/SamSaidYes.PostsService.Tests/ApiTest/Analytics/LeaderBoardServiceTest.cs:line 25
   at InvokeStub_LeaderBoardServiceTest..ctor(Object, Span`1)
   at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

as you can tell by the logs we're using xUnit (which I assume pretty popular in general).

``` xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <TargetFramework>net9.0</TargetFramework>
        <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
        <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
        <LangVersion>default</LangVersion>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="Bogus" Version="35.6.1"/>
        <PackageReference Include="FluentAssertions" Version="6.12.1"/>
        <PackageReference Include="FluentAssertions.Web" Version="1.5.0"/>
        <PackageReference Include="FluentAssertions.Web.Serializers.NewtonsoftJson" Version="1.5.0"/>
        <PackageReference Include="JetBrains.Annotations" Version="2023.3.0"/>
        <PackageReference Include="Microsoft.AspNet.WebApi.Client" Version="6.0.0"/>
        <PackageReference Include="Microsoft.AspNetCore.TestHost" Version="9.0.0-rc.2.24474.3"/>
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.11.1"/>
        <PackageReference Include="NSubstitute" Version="5.1.0"/>
        <PackageReference Include="RavenDB.TestDriver" Version="6.2.0"/>
        <PackageReference Include="Serilog.Sinks.XUnit" Version="3.0.5"/>
        <PackageReference Include="Snapshooter.Xunit" Version="0.14.1"/>
        <PackageReference Include="xunit" Version="2.9.0"/>
        <PackageReference Include="xunit.runner.visualstudio" Version="2.8.2">
            <PrivateAssets>all</PrivateAssets>
            <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
        </PackageReference>
    </ItemGroup>

</Project>

```

Is there an alternative way to deal with this?

 and could this please be at least documented somewhere or become the default the Tests because is really time consuming of it's non deterministic nature and IMHO this is not helping with the claim `the boring database` ?

**Comment by SbiCA** [comment]
Update after some analysis I figured out that at least one project had two bases classes configuring the same.

1. RavenIntegrationTest (inherits from RavenTestDriver), configured with Option 1) static constructor
2. BaseRavenTestDriver  (inherits from RavenTestDriver) ,  configured with Option 1) static constructor

this seems to lead to bad behaviour I'm not sure if it's due to the xunit runner or why this happens. 
Once I unified the base class the behaviour is back to "working" but this is still unexpected to me why is the test class driver causing this problem 

also I can confirm after downgrading to `6.0.106` it works so it's a recent change

**Comment by karmeli87** [comment, marked as answer]
Hi,

Indeed this change was introduced recently, you can read more about it [here](https://ravendb.net/docs/article-page/6.2/csharp/migration/embedded/testdriver-breaking-changes)

There is a concurrency check to prevent double configuration 
https://github.com/ravendb/ravendb/blob/3c20124c7801117b65cb76e344aca49a9de03bc7/src/Raven.TestDriver/RavenTestDriver.cs#L62-L63
and this is exactly the exception that you are getting.

You configure it as part of the static ctor, and it throws during that static ctor initialization, which will cause the following:
> If a static constructor throws an exception, the runtime doesn't invoke it a second time, and the type remains uninitialized for the lifetime of the application domain.

(source: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors)

**Reply by SbiCA** [reply]
@karmeli87  thx for adding the docs the thing that tripped us of was the different xUnit test classes which made it non deterministic based on where it would get called first. Nevertheless once we fixed our mess it works as described in the docs 👏  

I think if this causes head aches with other users, I would be in favour having the license check off by default for test because at least from my perspective I can't imagine writing a test that would require a license

**Reply by ppekrol** [reply]
We would prefer user to make that decision consciously, because when you run without licenses the server is running only on 3 processor cores. In many cases, adding a license might boost significantly the test execution speed.

**Reply by SbiCA** [reply]
Understood you're point I think my view is different or I'm assuming implicitly. Our typical CI test usually involve spinning up an isolated in memory environment using WAF + RavenTestDriver so that you get a fast isolation, this somehow limits the "size" of tests, hence the said benefit in our case won't be noticeable. On the other hand if I'd like to run dedicate perf tests than I agree with that statement.

To me I'm just considering 80/20 here that's why I'd still be for an opt-in for the license instead because we're writing rarely perf tests on that scale (As ordinary application/LOB devs).
